### PR TITLE
cpu/stm32: Fix & cleanup periph_eth

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

The function to check the link status of the STM32 Ethernet peripheral used the hard coded PHY address 0, rather than the configured address. This made me realize that the API of the MII register access functions sucks, as only one PHY is supported anyway.

- Dropped the PHY address parameter in the MII register read and write functions
    - Instead, the configured PHY address is used
- Updated all calls to it
- Fixed the PHY address of the Nucleos
    - The reason why address 0 worked for reading the link status is that apparently 0 is the correct PHY address
    - The Ethernet on Nucleo-F767ZI works as good as before with PHY address 0 (with occasionally the PHY getting configured to 10 Mbps half-duplex while the MAC is configured at 100 Mbps full-duplex, resulting in https://github.com/RIOT-OS/RIOT/issues/13490 that is already present in master)

### Testing procedure

E.g. flash `examples/gnrc_networking`
- `ifconfig` should still properly see the link status
- `ping6 <addr>` should work, unless the issue in https://github.com/RIOT-OS/RIOT/issues/13490 hits (roughly one out of three boots)

### Issues/PRs references

None

### CODE REVIEW

**Comment by maribu** [comment]
@JannesVolkens: You have a Nucleo-F207ZG if I recall correctly. Could you test this PR on this board?

@bergzand: You have a Nucleo-F746ZG if I recall correctly. Could you test this PR?

@aabadie: Care to confirm my test results on the Nucleo-F767ZI?

**Comment by maribu** [comment]
BTW: Now that I know what is causing https://github.com/RIOT-OS/RIOT/issues/13490 I should come up with a fix swiftly. But first let me split out the [MII](https://en.wikipedia.org/wiki/Media-independent_interface) definitions from `cpu/stm32/include/periph_cpu.h` to a common utility header. The MII register specification is a standard used by most (all?) Ethernet network interfaces - we should have this as common header. E.g. [Linux](https://elixir.bootlin.com/linux/v5.8.14/source/include/uapi/linux/mii.h) also has a common utility header for this, rather than reinventing the wheel for every Ethernet adapter.

**Comment by maribu** [comment]
Fixed typo in commit message

**Comment by JannesVolkens** [comment]
Everything is working correctly on the nucleo-f207zg!

Output:
`ifconfig` with with an active link:
```
ifconfig
2020-10-09 12:36:11,367 #  ifconfig
2020-10-09 12:36:11,371 # Iface  5  HWaddr: 26:1B:10:41:03:29  Link: up 
2020-10-09 12:36:11,375 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2020-10-09 12:36:11,377 #           RTR_ADV  
2020-10-09 12:36:11,380 #           Source address length: 6
2020-10-09 12:36:11,383 #           Link type: wired
2020-10-09 12:36:11,388 #           inet6 addr: fe80::241b:10ff:fe41:329  scope: link  VAL
2020-10-09 12:36:11,391 #           inet6 group: ff02::2
2020-10-09 12:36:11,394 #           inet6 group: ff02::1
2020-10-09 12:36:11,397 #           inet6 group: ff02::1:ff41:329
2020-10-09 12:36:11,398 #           
2020-10-09 12:36:11,401 #           Statistics for Layer 2
2020-10-09 12:36:11,404 #             RX packets 7  bytes 1084
2020-10-09 12:36:11,409 #             TX packets 1 (Multicast: 1)  bytes 78
2020-10-09 12:36:11,412 #             TX succeeded 1 errors 0
2020-10-09 12:36:11,415 #           Statistics for IPv6
2020-10-09 12:36:11,418 #             RX packets 2  bytes 192
2020-10-09 12:36:11,422 #             TX packets 1 (Multicast: 1)  bytes 64
2020-10-09 12:36:11,425 #             TX succeeded 1 errors 0
2020-10-09 12:36:11,425 # 
```

`ifconfig` without an active link:
```
ifconfig
2020-10-09 12:36:18,187 #  ifconfig
2020-10-09 12:36:18,191 # Iface  5  HWaddr: 26:1B:10:41:03:29  Link: down 
2020-10-09 12:36:18,195 #           L2-PDU:1500  MTU:1500  HL:64  RTR  
2020-10-09 12:36:18,197 #           RTR_ADV  
2020-10-09 12:36:18,200 #           Source address length: 6
2020-10-09 12:36:18,203 #           Link type: wired
2020-10-09 12:36:18,208 #           inet6 addr: fe80::241b:10ff:fe41:329  scope: link  VAL
2020-10-09 12:36:18,211 #           inet6 group: ff02::2
2020-10-09 12:36:18,214 #           inet6 group: ff02::1
2020-10-09 12:36:18,217 #           inet6 group: ff02::1:ff41:329
2020-10-09 12:36:18,218 #           
2020-10-09 12:36:18,221 #           Statistics for Layer 2
2020-10-09 12:36:18,224 #             RX packets 45  bytes 8782
2020-10-09 12:36:18,229 #             TX packets 4 (Multicast: 2)  bytes 328
2020-10-09 12:36:18,232 #             TX succeeded 4 errors 0
2020-10-09 12:36:18,235 #           Statistics for IPv6
2020-10-09 12:36:18,238 #             RX packets 39  bytes 7030
2020-10-09 12:36:18,242 #             TX packets 4 (Multicast: 2)  bytes 272
2020-10-09 12:36:18,245 #             TX succeeded 4 errors 0
2020-10-09 12:36:18,246 # 
```

`ping6` with an active link:
```
ping6 fe80::d941:e3fd:1e26:ad1b
2020-10-09 12:36:24,712 #  ping6 fe80::d941:e3fd:1e26:ad1b
2020-10-09 12:36:24,719 # 12 bytes from fe80::d941:e3fd:1e26:ad1b%5: icmp_seq=0 ttl=64 time=0.247 ms
2020-10-09 12:36:25,721 # 12 bytes from fe80::d941:e3fd:1e26:ad1b%5: icmp_seq=1 ttl=64 time=1.447 ms
2020-10-09 12:36:26,722 # 12 bytes from fe80::d941:e3fd:1e26:ad1b%5: icmp_seq=2 ttl=64 time=1.474 ms
2020-10-09 12:36:26,722 # 
2020-10-09 12:36:26,725 # --- fe80::d941:e3fd:1e26:ad1b PING statistics ---
2020-10-09 12:36:26,730 # 3 packets transmitted, 3 packets received, 0% packet loss
2020-10-09 12:36:26,734 # round-trip min/avg/max = 0.247/1.056/1.474 ms
```

`ping6` without an active link:
```
ping6 fe80::d941:e3fd:1e26:ad1b
2020-10-09 12:36:29,184 #  ping6 fe80::d941:e3fd:1e26:ad1b
2020-10-09 12:36:32,184 # 
2020-10-09 12:36:32,188 # --- fe80::d941:e3fd:1e26:ad1b PING statistics ---
2020-10-09 12:36:32,194 # 3 packets transmitted, 0 packets received, 100% packet loss
```

**Comment by benpicco** [comment]
Will this need a backport?

**Comment by maribu** [comment]
> Will this need a backport?

Wouldn't hurt :-)

**Comment by bergzand** [comment]
> Wouldn't hurt :-)

I agree that this fixes a bug. I'm fine with a backport for this. Thanks!

**Comment by maribu** [comment]
Hmm, I think I should use `uint16_t` for the phy addr - even though 0-31 comfortably fits into `uint8_t`. Bit shift and integers of different width is a can of worms I mighy better not open

**Comment by maribu** [comment]
> Hmm, I think I should use `uint16_t` for the phy addr - even though 0-31 comfortably fits into `uint8_t`. Bit shift and integers of different width is a can of worms I might better not open

Fixed. Test results remain valid: Due to [integer promotion](https://wiki.sei.cmu.edu/confluence/display/c/EXP14-C.+Beware+of+integer+promotion+when+performing+bitwise+operations+on+integer+types+smaller+than+int) of `uint8_t` the result of the operations is still identical. But with `uint16_t` used, I do no longer need to recall details of the C specification to explain why the code is correct :-)

> I agree that this fixes a bug. I'm fine with a backport for this. Thanks!

I propose to only backport the second commit and keep the hard coded phy address of 0 for the link status. All Nucleos supported by that branch use 0 as phy address (so the hard coded value is correct for all of them) and no new board support is going to be back-ported. Otherwise I would have to decide on whether to also backport the clean-up, or fix it completely differently in the stable branch.

File: cpu/stm32/periph/eth.c [hunk 1]
```diff
@@ -116,16 +115,16 @@ static uint8_t _link_state = LINK_STATE_DOWN;
...
-static uint16_t _mii_reg_transfer(unsigned addr, unsigned reg, uint16_t value,
-                                  bool write)
+static uint16_t _mii_reg_transfer(unsigned reg, uint16_t value, bool write)
 {
     unsigned tmp;
+    const uint16_t phy_addr = eth_config.phy_addr;
 
     while (ETH->MACMIIAR & ETH_MACMIIAR_MB) {}
-    DEBUG("[stm32_eth] rw_phy %x (%x): %x\n", addr, reg, value);
+    DEBUG("[stm32_eth] rw_phy %x (%x): %x\n", (unsigned)phy_addr, reg, value);
```
Comment (benpicco) [inline cpu/stm32/periph/eth.c:124]:
```suggestion
    DEBUG("[stm32_eth] rw_phy %x (%x): %x\n", phy_addr, reg, value);
```

you can drop the cast

**Review by benpicco — APPROVED** [review]
Pretty straightforward and @JannesVolkens confirmed it's working.

File: cpu/stm32/periph/eth.c (hunk 1 above)
Comment (maribu) [reply cpu/stm32/periph/eth.c:124]: `%x` expects `unsigned int` as type, `phy_addr` is `uint16_t`. (I am aware that for variadic functions integers smaller than `int` get promoted to `int`. But still the correct format specifier would technically be `%hx` and depending on compiler settings, you can get a warning about incorrect format specifier without the cast. I could also use `%hx` instead, but that would be even more obscure than the cast IMO.)

Comment (benpicco) [reply cpu/stm32/periph/eth.c:124]: I thought the compiler would only complain if the format specifier might truncate the value, not when it's 'too big'.

Comment (maribu) [reply cpu/stm32/periph/eth.c:124]: Note that variadic arguments have to go through the stack in any case, as they cannot easily be mapped to registers. E.g. using `%ld` on an `int` on a 64 bit machine is going to grab 32 bits more from the stack than it should.

**Comment by maribu** [comment]
Thanks!

**Pull request merged**